### PR TITLE
Github Task Lists

### PR DESCRIPTION
**Pull request by jhollingworth**
Add support for Github Task Lists under the gfm flag.

Changes to API
- list(_string_ body, _boolean_ ordered, _boolean_ taskList)
- listitem(_string_ text, [_boolean_ checked]).

`checked` is defined when you have a list item which starts with `[ ]` or
`[x]`.If defined its a boolean depending on whether the `x` is
present. When checked is defined we add a input type type `checkbox` to
the list item and add the class `task-list-item-checkbox`.

`taskList` is true if a list has any list items where `checked` is
defined. When true we add the class `task-list` to the list.

Resolves chjj/marked#107

### CODE REVIEW

**Comment by styfle** [comment]
@jhollingworth 
Very cool. Hopefully this gets merged.
However this project doesn't seem very active anymore.
You might want to checkout [markdown-it](https://github.com/markdown-it/markdown-it)

**Comment by chjj** [comment]
I had an implementation of this that I reverted because I was originally completely against further GFM extras. I thought I had lost it when my laptop bricked, but I realized it wasn't on a separate unpushed branch. It's in the git history: https://github.com/chjj/marked/issues/107

I'll compare these two.

**Comment by jhollingworth** [comment]
do you mean 26cef98d7261cb8330c07eceff250254124f20d8 and a5e39a6a59a91a7c8fa88e9113ba4dde7264e33b? The one issue I saw with yours was it does not support the scenario where a list has items that are not checkboxes. 

* [ ] Foo
* Bar
* [ ] Baz

Also, have your view changed on adding GFM extras or do you not want to merge this?

File: lib/marked.js [hunk 1]
```diff
@@ -27,6 +27,7 @@ var block = {
   text: /^[^\n]+/
 };
 
+block.checkbox = /^\[([ x])\] +/;
```
Comment (toolbear) [inline lib/marked.js]: According to GFM docs `[v] ` is a valid way to indicate a task is complete as well.

File: lib/marked.js [hunk 2]
```diff
@@ -304,6 +306,17 @@ Lexer.prototype.token = function(src, top, bq) {
         space = item.length;
         item = item.replace(/^ *([*+-]|\d+\.) +/, '');
 
+        if (this.options.gfm) {
+          checked = block.checkbox.exec(item);
+
+          if (checked) {
+            checked = checked[1] === 'x';
```
Comment (toolbear) [inline lib/marked.js]: Likewise, `x` or `v` can indicate a completed task.

**Comment by toolbear** [comment]
@jhollingworth thanks for starting with the lexer as that fits my use case.

**Comment by felixSchl** [comment]
What's happening with this?

**Comment by cdaguerre** [comment]
Will this make it to core? Any ETA?

**Comment by dvcrn** [comment]
Using this code in a app right now and seems to work good. Would love to see this getting merged back

screen shot 2015-11-11 at 2 39 31 pm

**Comment by msto** [comment]
Any chance of this getting pulled in in the near future? Would love to have checkboxes available in ipython notebooks, and they're waiting on it to be implemented here.

https://github.com/ipython/ipython/pull/8402
https://github.com/ipython/ipython/issues/6288

**Comment by noerw** [comment]
would love to see this too

**Comment by Sohalt** [comment]
Any updates on this?
This would make integration of checklists for https://github.com/wekan/wekan much easier.

**Comment by ousamabenyounes** [comment]
Checklist feature is an important feature for wekan project...
So merging this pull request would help developpers to finish this other pull request https://github.com/wekan/wekan/issues/298#issuecomment-181293631 and be sure they won't use another alternative...Hope to see this on wekan project soon...

**Comment by chrisdias** [comment]
Hi! Any chance this will get merged?  We would like top pick up this fix for [VS Code] (https://github.com/microsoft/vscode).

**Comment by dvcrn** [comment]
This project is definitely dead. I merged this in a own fork and use that for [markright](https://github.com/dvcrn/markright)

**Comment by ZackWard** [comment]
@matt- @parleur Thank you for maintaining this project. I would love to +1 this PR.

**Comment by Xaekai** [comment]
This should either be merged or feedback given as to why it won't be. What's the hold up?

**Comment by wujr5** [comment]
I hope this branch would be merged

**Comment by richarddavenport** [comment]
I write my notes in [VS Code](https://github.com/microsoft/vscode) using markdown. Every time I go to make a task list I expect GFM behavior, but alas, it never works. I'd love for VS Code to change, but I think they've made [it clear](https://github.com/Microsoft/vscode/issues/2175) that they won't. The state of this project is making me sad. Come on guys, get your act together, be a man, pass the baton and let people help.
@matt- @chjj @parleur

**Comment by matt-** [comment]
Not sure what "be a man" has to do with this PR. Please try to keep it professional.

**Comment by richarddavenport** [comment]
Thanks for approving the changes. May 7th 2015, was the last time any maintainer touched this issue. I was saying be a man about handing over the project to someone that could be more actively involved. For me it's a humbling experience to hand over a project that I started and grew. If you guys are not going to listening to the community, then that is unprofessional. Sometimes people need a little prodding.

**Comment by matt-** [comment]
What does being a man have to do with anything? What about ones sex do you believe solves this PR?  Grow up a bit man. You don't get thing done by being a dick. Fork the project and move on if you are so unhappy. That's the great thing about open source.

**Comment by matt-** [comment]
No hurt feelings. Just asking for some professionalism. Maybe too big of an ask, but can we at least get back on the topic of this ticket and less about your general complaints.

**Comment by richarddavenport** [comment]
I agree! Sorry I took out some aggression on this issue and the maintainers. Just really want my checkboxes in VS Code!

**Comment by styfle** [comment]
@matt- Good to see you approve this PR. What are the next steps to get this merged? Thanks!

**Comment by samiraguiar** [comment]
Any plans on getting this merged?

**Comment by Agent01** [comment]
I'd also love to see this merged!

**Comment by mspiegel** [comment]
+1

**Comment by jpouellet** [comment]
+1

I want this for https://github.com/volca/markdown-preview/issues/34

**Comment by bouzlibop** [comment]
+1

**Comment by omarzion** [comment]
I too am hoping this will be merged, in the meantime I am using a simple wrapper function to achieve this.  Maybe this will help someone until then.

```
function markdownFormatter(text) {
    var a = marked(text, {break: true});
    a = a.replace(/<a/g, '<a target="_blank"')
         .replace(/<li>\[\s\]/g, '<li class="check-box"><input onclick="return false" type="checkbox">')
         .replace(/<li>\[x\]/g, '<li class="check-box"><input checked onclick="return false" type="checkbox">');
    return a;
  }
```

**Comment by jbddc** [comment]
+1

**Comment by stanleykylee** [comment]
@chjj @matt- any updates on whether or not this will be merged as it is approved now?

**Comment by heaven00** [comment]
would love to have this!

**Comment by styfle** [comment]
To [reference](https://github.com/chjj/marked/pull/587#issuecomment-100055513) myself above:

Many users have migrated from `marked` to `markdown-it` due to inactivity and failure to merge the 163 pending PRs. This PR in particular has at least 30 people interested but it's mostly too late.

@chrisdias @richarddavenport
VS Code switched to `markdown-it` and now has plugin to support Task Lists
https://github.com/mjbvz/vscode-markdown-checkboxes

**Comment by samiraguiar** [comment]
Thank you, @styfle, I had missed your comment. Now I can finally unsubscribe from this thread.

**Comment by Xaekai** [comment]
I guess I'm going to give up too and switch to `markdown-it` 

The negligence of the maintainers here is unacceptable. I've waited over a year.

**Comment by joshbruce** [comment]
See #956 

Focusing on CommonMark and GFM specifications, without extensions. Cleaning house and assessing the damage. If this gets implemented it will probably be through extension via open/closed.

Thanks, and apologies for the inconvenience.

**Comment by joshbruce** [comment]
Reopening as things have changed a bit. We are looking to support GFM, which is only CommonMark with extensions. We have two test beds at present one for CommonMark the other for GFM. There are two failing tests (both examples) in the task list test suite. 

Can we verify this solution causes those to pass?

File: README.md [hunk 3]
```diff
@@ -211,7 +212,8 @@ This code will output the following HTML:
 - codespan(*string* code)
 - br()
 - del(*string* text)
-- link(*string* href, *string* title, *string* text)
+- link(*string* href, *string* title, *string* text, [*boolean* checked]).
+  - `checked` only defined when `gfm` is `true` and there is a check box at the start of the list item (e.g. `* [ ] foo`).
```
Comment (styfle) [inline README.md]: This file was moved to [USING_PRO.md](https://github.com/markedjs/marked/blob/master/docs/USING_PRO.md). and is causing merge conflicts.

**Comment by styfle** [comment]
@jhollingworth Are you still interested in working on this PR?

**Comment by joshbruce** [comment]
Going to go ahead and close as overcome by events. If you would like to continue working toward this goal:

1. Read #1216 
2. Submit a PR that solves GFM example 272 and 273 in the `gfm/gfm-spec.js` file.

**Comment by 30Wedge** [comment]
*Ping*; I want to show that I'm still interested in this PR. 
Jupyter uses marked, and I'd like to use gfm task lists in my notebook.

**Comment by joshbruce** [comment]
@30Wedge: What version of Marked are you using?

0.4.0 should include task lists if I'm remembering #1250 properly.

**Comment by 30Wedge** [comment]
Looks like they were using 0.3.0
Thanks for responding so quickly!

**Pull request closed without merging**